File: lib_recyclerviewadapter/src/main/java/lib/kalu/recyclerview/model/TransModel.java
package lib.kalu.recyclerview.model;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

/**
 * description: 多级菜单
 * created by kalu on 2017/5/26 15:11
 */
public abstract class TransModel<T> implements Serializable {

    // 默认合并
    protected boolean isExpand = false;
    // 数据集合
    protected List<T> modelList;

    /*****************************************************************************************/

    public abstract int getLevel();

    public boolean isExpanded() {
        return isExpand;
    }

    public void setExpanded(boolean expanded) {
        isExpand = expanded;
    }

    public List<T> getModelList() {
        return modelList;
    }

    public void setModelList(List<T> modelList) {
        this.modelList = modelList;
    }

    /*****************************************************************************************/

    public boolean hasModel() {
        return modelList != null && modelList.size() > 0;
    }

    public T getModel(int position) {
        if (hasModel() && position < modelList.size()) {
            return modelList.get(position);
        } else {
            return null;
        }
    }

    public int getModelPosition(T subItem) {
        return modelList != null ? modelList.indexOf(subItem) : -1;
    }

    public void addModel(T subItem) {
        if (modelList == null) {
            modelList = new ArrayList<>();
        }
        modelList.add(subItem);
    }

    public void addModel(int position, T subItem) {
        if (modelList != null && position >= 0 && position < modelList.size()) {
            modelList.add(position, subItem);
        } else {
            addModel(subItem);
        }
    }

    public boolean contains(T subItem) {
        return modelList != null && modelList.contains(subItem);
    }

    public boolean removeModel(T subItem) {
        return modelList != null && modelList.remove(subItem);
    }

    public boolean removeModel(int position) {
        if (modelList != null && position >= 0 && position < modelList.size()) {
            modelList.remove(position);
            return true;
        }
        return false;
    }
}

File: lib_kotlin/src/main/java/lib/kalu/adapter/model/TransModel.kt
package lib.kalu.adapter.model

import java.util.ArrayList

/**
 * description: 多级菜单
 * created by kalu on 2017/5/26 15:11
 */
abstract class TransModel<T> {

    // 默认合并
    var isExpanded = false
    // 数据集合
    private var modelList: MutableList<T>? = null

    abstract val level: Int

    fun getModelList(): MutableList<T>? {
        return modelList
    }

    fun setModelList(modelList: MutableList<T>) {
        this.modelList = modelList
    }

    /** */

    fun hasModel(): Boolean {
        return modelList != null && modelList!!.size > 0
    }

    fun getModel(position: Int): T? {
        return if (hasModel() && position < modelList!!.size) {
            modelList!![position]
        } else {
            null
        }
    }

    fun getModelPosition(subItem: T): Int {
        return if (modelList != null) modelList!!.indexOf(subItem) else -1
    }

    fun addModel(subItem: T) {
        if (modelList == null) {
            modelList = ArrayList()
        }
        modelList!!.add(subItem)
    }

    fun addModel(position: Int, subItem: T) {
        if (modelList != null && position >= 0 && position < modelList!!.size) {
            modelList!!.add(position, subItem)
        } else {
            addModel(subItem)
        }
    }

    operator fun contains(subItem: T): Boolean {
        return modelList != null && modelList!!.contains(subItem)
    }

    fun removeModel(subItem: T): Boolean {
        return modelList != null && modelList!!.remove(subItem)
    }

    fun removeModel(position: Int): Boolean {
        if (modelList != null && position >= 0 && position < modelList!!.size) {
            modelList!!.removeAt(position)
            return true
        }
        return false
    }
}
